<a href="https://colab.research.google.com/github/kennylin8212/tibame1224-1/blob/main/%E6%B8%AC%E8%A9%A6%E7%84%A1%E7%AA%AE%E8%BF%B4%E5%9C%88%E7%AD%89%E7%9B%B8%E9%97%9C%E9%8C%AF%E8%AA%A4%E7%A4%BA%E7%AF%84.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#自行輸入特定日期(today),往後數10個完全平方日
def is_leap_year(year):
    if year % 400 == 0:
        return True
    elif year % 100 == 0:
        return False
    elif year % 4 == 0:
        return True
    else:
        return False

def is_valid_date(n):
    year = n // 10000
    last4 = n % 10000
    day = last4 % 100
    month = last4 // 100
    # print(year, month, day)

    max_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    if 1 <= month <= 12:
        month_day = max_days[month-1]
        # 考慮閏年
        if month == 2 and is_leap_year(year) == True:
            month_day = month_day + 1

        if 1 <= day <= month_day:
            # print("合理")
            return True
        else:
            # print("不合理日期(天)")
            return False
    else:
        # print("不合理日期(月)")
        return False

import datetime
today = int(input("請輸入日期:"))
y = today // 10000
m = today % 10000 // 100
d = today % 100
count = 0

if is_valid_date(today) == True:
    while True:
        start = datetime.datetime(y,m,d)
        start = start + datetime.timedelta(days=1)
        #start只是一個 datetime 對象(物件),並非數字也不是字串型態
        #strftime 將 datetime 對象轉換為特定格式的字串
        #後續math.isqrt() 需要操作整數,故轉換成int
        date_as_int = int(start.strftime("%Y%m%d"))
        import math
        sqrt_today = math.isqrt(date_as_int)
        if sqrt_today * sqrt_today == date_as_int:
          print(start.strftime("%Y/%m/%d"))
          count = count +1
        if count == 10:
            break
        else:
            print("異常")
            break
else:
    print("您輸入的日期不合理")

請輸入日期:20241231
異常


        start = datetime.datetime(y,m,d)20241231
        start = start + datetime.timedelta(days=1)20250101  
        迴圈中兩個start放在一起會衝突形成無窮迴圈
        
        本次迴圈+1後下次再進迴圈又還原再+1 還是永遠停留在20250101
        count 永遠不會等於 10 永遠不會跳出迴圈
        若本例沒有 else 判斷異常的話,就會線如無窮迴圈沒答案
        
        解決方法: 應把第一個start(類似totoal = 0 )放在迴圈外


In [6]:
#若沒有把第一個a放外面,跑十次迴圈都不會累加,始終留在1
for i in range(10):
    a = 0
    a= a + 1
print(a)
#十次迴圈跑完才印

1


In [7]:
for i in range(10):
    a = 0
    a= a + 1
    print(a)
#每一次迴圈都印

1
1
1
1
1
1
1
1
1
1
